### PR TITLE
Derivatives of relu (#63027)

### LINKED ISSUE #63027: ReLU uses more VRAM for backward-pass than needed
## 🚀 Feature

Storing the ReLU output Tensor for backward-pass instead of the input Tensor.

Both the input- and output-Tensor should be equivalent regarding the gradient calculation for ReLUs. However, the output-Tensor is usually already stored for calculation of the backward-pass for the following module, giving an opportunity to reduce VRAM allocated by the GPU.

## Motivation

Large models use a lot of VRAM, often reach the limit of a GPU, and ReLU activation functions may be used several times within a network, resulting in a large memory saving potential.

## Pitch

1. Instead of saving the input Tensor for the torch.nn.ReLU backward-pass, the `output = th.relu(input)` of the module may be saved for the backward-pass.
2. During the backward-pass, the input Tensor is replaced by the output Tensor, e.g. `grad *= output>0`, or however this is realized in the PyTorch code.

This should be applicable due to output-values being <=0 for both the input and the output Tensors.

This saves VRAM, due to ReLU outputs being often an input to following modules, where the input is usually saved for backward-pass gradient calculations. Thus, the output Tensor (ReLU) has only to be saved once, whereas the input Tensor (ReLU) needs to be saved twice.

## Alternatives

Leaving torch.nn.ReLU as is.

## Additional context

The following script repeatedly applies the same Conv2d module and either the new ReLU or torch.nn.ReLU in a closed-loop set-up, and calculates the gradients during a backward-pass.

The script shows an improvement for ReLU regarding VRAM usage (without model parameter, input, label), and an exemplary equivalence of gradients for both versions.

However, the forward+backward pass duration is dependent on which Module is evaluated first, resulting in the same runtime when running the new ReLU first, and in a faster runtime for the new ReLU when running torch.nn.ReLU first.

```
import torch as th
from torch import nn
from torch.autograd import Function

import time

ITERATION = 10
DEVICE = 0
SHAPE = [1, 256, 512, 512]


class ReluFunction(Function):
    @staticmethod
    def forward(ctx, input: th.Tensor) -> th.Tensor:
        relu = th.relu(input)
        ctx.save_for_backward(relu)
        return relu

    @staticmethod
    def backward(ctx, grad: th.Tensor):
        relu, = ctx.saved_tensors
        grad *= (relu > 0)
        return grad


class ReLU(nn.Module):
    def __init__(self):
        super(ReLU, self).__init__()

    def forward(self, input: th.Tensor) -> th.Tensor:
        return ReluFunction.apply(input)


def speed_test(model: nn.Module, module: nn.Module, input: th.Tensor, output: th.Tensor, criterion, iteration=10):
    device = input.device
    module = module.to(device=device)

    input.grad = None
    input.requires_grad_()

    model = model.to(device=device)

    th.cuda.synchronize(device)
    th.cuda.reset_peak_memory_stats(device)
    vram_0 = th.cuda.max_memory_allocated(device)
    time_0 = time.time()

    prediction = input
    for _ in range(iteration):
        prediction = model(prediction)
        prediction = module(prediction)

    th.cuda.synchronize(device)
    vram_forward = th.cuda.max_memory_allocated(device)
    loss = criterion(prediction, output)
    loss.backward()

    th.cuda.synchronize(device)
    print(f'Forward + Backward: {time.time() - time_0:1.2f}s, \n'
          f'Forward (VRAM):     {(vram_forward - vram_0)/1e9:1.2f}GB, \n'
          f'Backward (VRAM):    {(th.cuda.max_memory_allocated(device) - vram_forward) / 1e9:1.2f}GB')

    return prediction, input.grad


if __name__ == "__main__":
    device = th.device(DEVICE)

    input = th.rand(size=SHAPE, device=device)
    label = th.rand_like(input)

    model = nn.Conv2d(
        in_channels=input.shape[1],
        out_channels=input.shape[1],
        kernel_size=(3, 3),
        padding=(1, 1),
        bias=True).to(device=device)

    criterion = nn.MSELoss()

    # For initializing device
    criterion(model(input), label).backward()

    print('New ReLU:')
    pr_new = speed_test(model, ReLU(), input, label, criterion, ITERATION)
    print('ReLU:')
    pr_relu = speed_test(model, nn.ReLU(), input, label, criterion, ITERATION)

    print('')
    print(f'Max abs difference prediction: {th.max((pr_new[0] - pr_relu[0]).abs()).detach().cpu():1.2e}')
    print(f'Max abs difference grad:       {th.max((pr_new[1] - pr_relu[1]).abs()).detach().cpu():1.2e}')

```

### Output:
New ReLU:
Forward + Backward: 0.50s, 
Forward (VRAM):     2.95GB, 
Backward (VRAM):    1.49GB

ReLU:
Forward + Backward: 0.50s, 
Forward (VRAM):     5.37GB, 
Backward (VRAM):    1.49GB

Max abs difference prediction: 0.00e+00
Max abs difference grad:       0.00e+00

### PR DESCRIPTION
**Pull request by MengeTM**
Optimization of relu and leaky_relu derivatives for reduction of VRAM needed for the backward-passes

Fixes #63027

### CODE REVIEW

File: tools/autograd/derivatives.yaml [hunk 1]
```diff
@@ -1598,7 +1598,7 @@
   self: soft_margin_loss_backward(grad, self, target, reduction)
 
 - name: relu(Tensor self) -> Tensor
-  self: threshold_backward(grad, self, 0)
+  self: threshold_backward(grad, result, 0)
```
Comment (albanD) [inline tools/autograd/derivatives.yaml:1601]: You can remove the inplace version below for both.
The non-inplace version will be used for both automatically.

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: When removing the inplace versions, test_leaky_relu_inplace_with_neg_slope_... result in an AssertionError "RuntimeError not raised"

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: It seems that we would need for LeakyReLU to use different backward algorithms regarding its negative_slope, which did allow values < 0 for leaky_relu, but not for leaky_relu, which results from using "result" for gradients.

A solution would be to save "result" for positive, and "self" for negative values, if it is doable with PyTorch

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: It seems that we would need for LeakyReLU to use different backward algorithms regarding its negative_slope, which did allow values < 0 for leaky_relu, but not for leaky_relu_, which results from using "result" for gradients.

A solution would be to save "result" for positive, and "self" for negative values - however it is done in PyTorch

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: It seems that we would need for LeakyReLU to use different backward algorithms regarding its negative_slope, which did allow values < 0 for leaky_relu, but not for leaky_relu_, which results from using "result" for gradients.

A solution would be to save "result" for positive, and "self" for negative values:
```yaml
self: "negative_slope >= 0 ? leaky_relu_backward(grad, result, negative_slope, false) : leaky_relu_backward(grad, self, negative_slope, false)"
```

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: It seems that c10::Scalar does not allow >=, thus we may have to add those to Scalar.cpp - how should I test if it actualy does what it should?

Comment (albanD) [reply tools/autograd/derivatives.yaml:1601]: Ho sorry, I forgot leaky relu doesn't have the same formula for inplace. My bad.
You can remove only relu_ then, thanks!

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: It should not work for leaky_relu as well when we replace "self" with "return".

Using "return" is only valid as a replacement for "self" when negative_slope value is positive, because the output has to behave the same regarding the 0 threshold as the input - with a negative_slope < 0 the output will be positive for self > 0 and self < 0, so that "result" becomes indistinguishable regarding the threshold, which would result in wrong gradients.

Thus, using the less VRAM needy version for leaky_relu will make its behavior equivalent to reaky_relu_, so that it should throw a Not Supported Error for negative_slope < 0 as well. However, this would change its behavior, which may result in errors for users. I recommend to either not improve the VRAM need for leaky_relu, which is up to 20% savings for a Conv2d net I use, or we should distinguish regarding the negative_slope and use "self" for negative negative_slopes

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: Should we remove the error message from Activation.cpp (line 116)?

Comment (albanD) [reply tools/autograd/derivatives.yaml:1601]: Yes, you can add back the inplace leaky relu version.

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: @albanD It seems that both "result" and "self" will be stored for LeakyReLU for the above solution, which are set at compile time. How may someone store backpropagation variables at run time, or how should we handle a negative_slope < 1 other than using the old leaky_relu or not allowing a negative_slope for the function?

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: If we add back the inplace leaky relu version, we will need to use "self" for the other relu version

Comment (MengeTM) [reply tools/autograd/derivatives.yaml:1601]: If we add back the inplace leaky relu version, we will need to use "self" for the leaky relu

Comment (albanD) [reply tools/autograd/derivatives.yaml:1601]: Sorry for the delay.
From checking the code, I think we want:
- Keep the current change to relu of always replacing self by result. It means that we can remove the inplace entry as the inplace and out of place now have the same formula.
- Leave leaky relu as it is today in master.
  - Using result in the out of place version would mean that we cannot  compute the gradient for negative slopes.
  - We could add a system to choose what is going to be saved based on the value of negative_slope. But that would be another feature, outside of the scope of the current PR.

**Review by albanD — APPROVED** [review]
Thanks!
